### ISSUE TITLE
Unable to sSet `use_regex=False` in BPE decoder & post_processor?

### ISSUE THREAD

**Issue by jchwenger** [huggingface/tokenizers#1563]
Hi there,

I'm attempting to create a byte-level BPE tokenizer that would be identical to GPT-2 except for the use of their original regex (for languages without spaces, and other use cases). I have seen threads by other users, but something caught my eye w.r.t. the decoder and post-processor. Here's my pipeline now:

```python
from tokenizers import Tokenizer
from tokenizers import AddedToken
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer

from tokenizers import pre_tokenizers
from tokenizers import processors
from tokenizers import decoders

from transformers import PreTrainedTokenizerFast

tokenizer = Tokenizer(BPE())

# Pre-Tokenizer: https://huggingface.co/docs/tokenizers/v0.13.4.rc2/en/api/pre-tokenizers#tokenizers.pre_tokenizers.ByteLevel
tokenizer.pre_tokenizer = pre_tokenizers.ByteLevel(
    add_prefix_space = False,
    use_regex = False
    # trim_offsets = True, # not available as arg but present in json?
)

# Processor: https://huggingface.co/docs/tokenizers/en/api/post-processors#tokenizers.processors.ByteLevel
tokenizer.post_processor = processors.ByteLevel(
    trim_offsets = False,
    # add_prefix_space = True, # not available as arg but both present in json?
    # use_regex = False 
)

# Decoder: https://huggingface.co/docs/tokenizers/en/api/decoders#tokenizers.decoders.ByteLevel
tokenizer.decoder = decoders.ByteLevel(
    # trim_offsets = True,     # not available as arg but all three present in json?
    # add_prefix_space = True,
    # use_regex = False 
)

special = AddedToken("<|endoftext|>", normalized = True)
trainer = BpeTrainer(vocab_size = 50257, special_tokens = [special])

tokenizer.train("data/dataset.txt", trainer = trainer)

tokenizer.add_special_tokens([special]) # is there a way to add the special token at the end of the vocabulary?

print(f"saving tokenizer to: {os.path.join(save_directory, tokenizer_name, 'tokenizer.json')}")
# tokenizer.save(os.path.join(save_directory, tokenizer_name, 'tokenizer.json')) # Save the tokenizer
tok_fast = PreTrainedTokenizerFast(
    tokenizer_object=tokenizer,
    model_max_length=1024,
    bos_token="<|endoftext|>",
    eos_token="<|endoftext|>",
    unk_token="<|endoftext|>",
)
tok_fast.save_pretrained(os.path.join(save_directory, tokenizer_name)) # Save the tokenizer
```

It seems I'm close to my goal, as the plain GPT-2 tokenizer, when using the autoclass, yields almost the same result:

```python
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained('gpt2',use_fast=True) 
tokenizer.save_pretrained(os.path.join(save_directory, tokenizer_name))
```

These will then yield the following json configs:

|self-made | GPT-2|
| --- | --- |
| <pre>{<br>  "version": "1.0",<br>  "truncation": null,<br>  "padding": null,<br>  "added_tokens": [<br>    {<br>      "id": 0,<br>      "content": "<\|endoftext\|>",<br>      "single_word": false,<br>      "lstrip": false,<br>      "rstrip": false,<br>      "normalized": true,<br>      "special": true<br>    }<br>  ],<br>  "normalizer": null,<br>  "pre_tokenizer": {<br>    "type": "ByteLevel",<br>    "add_prefix_space": false,<br>    "trim_offsets": true,<br>    "use_regex": false             # great! <br>  },<br>  "post_processor": {<br>    "type": "ByteLevel",<br>    "add_prefix_space": true,<br>    "trim_offsets": false,<br>    "use_regex": true              # why?! <br>  },<br>  "decoder": {<br>    "type": "ByteLevel",<br>    "add_prefix_space": true,<br>    "trim_offsets": true,<br>    "use_regex": true              # why?! <br>  },<br>  "model": {<br>    "type": "BPE",<br>    "dropout": null,<br>    "unk_token": null,<br>    "continuing_subword_prefix": null,<br>    "end_of_word_suffix": null,<br>    "fuse_unk": false,<br>    "byte_fallback": false,<br>    "ignore_merges": false,<br>    "vocab": {<br>      "<\|endoftext\|>": 0,<br>      ...<br>    },<br>    "merges": [<br>    ...<br>      ]<br>   }<br>}<br></pre>|<pre>{<br>  "version": "1.0",<br>  "truncation": null,<br>  "padding": null,<br>  "added_tokens": [<br>    {<br>      "id": 50256,<br>      "content": "<\|endoftext\|>",<br>      "single_word": false,<br>      "lstrip": false,<br>      "rstrip": false,<br>      "normalized": true,<br>      "special": true<br>    }<br>  ],<br>  "normalizer": null,<br>  "pre_tokenizer": {<br>    "type": "ByteLevel",<br>    "add_prefix_space": false,<br>    "trim_offsets": true,<br>    "use_regex": true<br>  },<br>  "post_processor": {<br>    "type": "ByteLevel",<br>    "add_prefix_space": true,<br>    "trim_offsets": false,<br>    "use_regex": true<br>  },<br>  "decoder": {<br>    "type": "ByteLevel",<br>    "add_prefix_space": true,<br>    "trim_offsets": true,<br>    "use_regex": true<br>  },<br>  "model": {<br>    "type": "BPE",<br>    "dropout": null,<br>    "unk_token": null,<br>    "continuing_subword_prefix": "",<br>    "end_of_word_suffix": "",<br>    "fuse_unk": false,<br>    "byte_fallback": false,<br>    "ignore_merges": false,<br>    "vocab": {<br>    ...<br>      "<\|endoftext\|>": 50256<br>    },<br>    "merges": [<br>    ...<br>    ]<br>  }<br>}<br></pre>|

Hence my question: what would be the correct way to make sure both the post-processor and the decoder have their `use_regex` set to False? Is it normal that the saved `.json` has these set to True in this set-up?

(Optional, not particularly important: does HF tokenizer have an option for setting the special token *after* the rest of the vocabulary? Somehow I could not find that information.)

Any help greatly appreciated, many thanks in advance!

tokenizers==0.19.1

**Comment by ArthurZucker** [comment]
Ah that is a good questions, I think there is a serialization bug let me have alook!

**Comment by ArthurZucker** [comment]
Okay, sorry it's counter intuitive but it is not used:
```rust
/// As a `Decoder`, `ByteLevel` is in charge of converting any byte-level characters to their
/// unicode counterpart, before merging everything back into a single String.
/// This decoder will consume the tokens and merge them in one step to alleviate
/// the fact that single token decoded might be a byte not representable as
/// as String.
```
Bytelevel is both and pre tokenizer and a decoder, but the decoder does not us the regex at all

**Comment by ArthurZucker** [comment]
(same for post-processor) you can safely ignore that!

**Comment by jchwenger** [comment]
Thanks @ArthurZucker for this, that's great!
